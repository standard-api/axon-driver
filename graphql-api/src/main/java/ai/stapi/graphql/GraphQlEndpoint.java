package ai.stapi.graphql;

import java.util.LinkedHashMap;
import java.util.Map;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.RestController;


@RestController
public class GraphQlEndpoint {

  private final GraphQlExecutor graphQlExecutor;

  public GraphQlEndpoint(
      GraphQlExecutor graphQlExecutor
  ) {
    this.graphQlExecutor = graphQlExecutor;
  }

  @PostMapping("/graphql")
  @ResponseBody
  public Map<String, Object> graphQl(@RequestBody GraphQlOperation graphQLOperation) {
    var response = new LinkedHashMap<String, Object>();
    response.put("data", this.graphQlExecutor.execute(graphQLOperation));
    return response;
  }
}
