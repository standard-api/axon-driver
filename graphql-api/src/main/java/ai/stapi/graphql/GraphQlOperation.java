package ai.stapi.graphql;

import java.util.HashMap;
import java.util.Map;

public class GraphQlOperation {

  private String operationName;
  private String query;

  private Map<String, Object> variables;

  protected GraphQlOperation() {
  }

  public GraphQlOperation(String operationName, String query, Map<String, Object> variables) {
    this.operationName = operationName;
    this.query = query;
    this.variables = variables;
  }

  public GraphQlOperation(String operationName, String query) {
    this(operationName, query, new HashMap<>());
  }

  public String getQuery() {
    return query;
  }

  public String getOperationName() {
    return operationName;
  }

  public Map<String, Object> getVariables() {
    return variables;
  }
}
