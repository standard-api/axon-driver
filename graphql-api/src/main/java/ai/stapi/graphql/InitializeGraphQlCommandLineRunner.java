package ai.stapi.graphql;

import ai.stapi.graphql.graphqlJava.graphQLProvider.GraphQLProvider;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.CommandLineRunner;

public class InitializeGraphQlCommandLineRunner implements CommandLineRunner {

  private final GraphQLProvider graphQLProvider;

  private final Logger logger;

  public InitializeGraphQlCommandLineRunner(
      GraphQLProvider graphQLProvider
  ) {
    this.graphQLProvider = graphQLProvider;
    this.logger = LoggerFactory.getLogger(InitializeGraphQlCommandLineRunner.class);
  }

  @Override
  public void run(String... args) {
    try {
      this.graphQLProvider.initialize();
    } catch (Exception exception) {
      this.logger.error("unable to generate graphql schema", exception);
    }
  }
}
