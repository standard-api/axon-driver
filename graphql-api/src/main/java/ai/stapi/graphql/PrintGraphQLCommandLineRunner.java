package ai.stapi.graphql;

import ai.stapi.graphql.generateGraphQlSchema.PrintGraphQlSchema;
import java.util.Arrays;
import org.axonframework.commandhandling.gateway.CommandGateway;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.context.ApplicationContext;


public class PrintGraphQLCommandLineRunner implements CommandLineRunner {

  private final CommandGateway commandGateway;

  private final ApplicationContext applicationContext;

  public PrintGraphQLCommandLineRunner(
      CommandGateway commandGateway,
      ApplicationContext applicationContext
  ) {
    this.commandGateway = commandGateway;
    this.applicationContext = applicationContext;
  }

  @Override
  public void run(String... args) {
    if (args.length == 0) {
      throw new RuntimeException("Please specify path where to print schema as command line argument.");
    }
    var path = Arrays.stream(args)
        .filter(arg -> arg.startsWith("_schemaOutputPath:"))
        .map(arg -> arg.replace("_schemaOutputPath:", ""))
        .findFirst();

    var command = new PrintGraphQlSchema(path.orElse(args[0]));
    this.commandGateway.sendAndWait(command);
    var exitCode = SpringApplication.exit(this.applicationContext, () -> 0);
    System.exit(exitCode);
  }
}
