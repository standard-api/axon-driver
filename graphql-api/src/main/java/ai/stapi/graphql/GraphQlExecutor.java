package ai.stapi.graphql;

import ai.stapi.graphql.graphqlJava.graphQLProvider.GraphQLProvider;
import graphql.ExecutionInput;
import java.util.Map;
import org.springframework.stereotype.Service;

@Service
public class GraphQlExecutor {

  private final GraphQLProvider schemaProvider;

  public GraphQlExecutor(GraphQLProvider schemaProvider) {
    this.schemaProvider = schemaProvider;
  }

  public Map<Object, Object> execute(GraphQlOperation graphQLOperation) {
    var executionInputBuilder = new ExecutionInput.Builder().query(graphQLOperation.getQuery());

    if (graphQLOperation.getOperationName() != null) {
      executionInputBuilder.operationName(graphQLOperation.getOperationName());
    }
    if (graphQLOperation.getVariables() != null) {
      executionInputBuilder.variables(graphQLOperation.getVariables());
    }
    var executionResult = this.schemaProvider.getGraphQL().execute(executionInputBuilder);
    return executionResult.getData();
  }

}
