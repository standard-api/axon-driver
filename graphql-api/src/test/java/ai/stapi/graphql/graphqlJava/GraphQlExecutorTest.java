package ai.stapi.graphql.graphqlJava;

import ai.stapi.graph.repositorypruner.RepositoryPruner;
import ai.stapi.graphoperations.synchronization.GraphSynchronizer;
import ai.stapi.graphql.GraphQlExecutor;
import ai.stapi.graphql.GraphQlOperation;
import ai.stapi.graphql.graphqlJava.testfixtures.ShowcaseGraphFixturesProvider;
import ai.stapi.graphql.graphqlJava.testfixtures.TestGraphqlModelDefinitionsLoader;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import ai.stapi.test.systemschema.SystemSchemaIntegrationTestCase;
import java.util.HashMap;
import java.util.Map;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

@StructureDefinitionScope(TestGraphqlModelDefinitionsLoader.SCOPE)
class GraphQlExecutorTest extends SystemSchemaIntegrationTestCase {

  @Autowired
  private GraphQlExecutor graphQlExecutor;

  @BeforeAll
  @Order(2)
  public static void beforeAll(
      @Autowired GraphSynchronizer graphSynchronizer,
      @Autowired RepositoryPruner repositoryPruner,
      @Autowired ShowcaseGraphFixturesProvider showcaseGraphFixturesProvider
  ) {
    repositoryPruner.prune();
    graphSynchronizer.synchronize(showcaseGraphFixturesProvider.getFixtureGraph());
  }

  @Test
  void itCanGetStructureDefinition() {
    var graphQlRequest = new GraphQlOperation(
        "",
        "query { getStructureDefinition(id: \"CodeableConcept\") { id kind abstract } }"
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanFindStructureDefinitions() {
    var graphQlRequest = new GraphQlOperation(
        "",
        "query { findStructureDefinitionList(sort: [ { type: ASC } ]) { id kind abstract } }"
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanGetStructureDefinitionAndItsElements() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    getStructureDefinition(id: "CodeableConcept") {
                        id
                        kind
                        abstract
                        differential {
                            element(
                                pagination: { limit: 100, offset: 0 },
                                sort: [
                                    { path: ASC }
                                ]
                            ) {
                                id
                                path
                                min
                                max
                            }
                        }
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanFindStructureDefinitionsAndGetOneAtTheSameTime() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findStructureDefinitionList(sort: [ { type: ASC } ]) {
                        id
                        kind
                        abstract
                    }
                    getStructureDefinition(id: "string") {
                        url
                        status
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanFindResourceWhichIsOnlyDefinedByStructureDefinition() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(sort: [ { name: ASC } ]) {
                        id
                        name
                        type {
                            coding {
                                system
                                code
                            }
                        }
                        address(sort: [ { city: ASC } ]) {
                            city
                            state
                            postalCode
                            country
                        }
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanSortByScalarFields() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(sort: [ { name: ASC } ]) {
                        id
                        name
                        type {
                            coding {
                                system
                                code
                            }
                        }
                        address(sort: [ { city: ASC } ]) {
                            city
                            state
                            postalCode
                            country
                        }
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanPaginate() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(sort: [ { name: ASC } ], pagination: { offset: 0, limit: 2 }) {
                        id
                        name
                        type {
                            coding {
                                system
                                code
                            }
                        }
                        address {
                            city
                            state
                            postalCode
                            country
                        }
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanSortByObjectFields() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(sort: [ { address: { city: DESC } } ]) {
                        id
                        name
                        type {
                            coding {
                                system
                                code
                            }
                        }
                        address(sort: [ { city: ASC } ]) {
                            city
                            state
                            postalCode
                            country
                        }
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanSortByEventDeeperObjectField() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(sort: [ { type: {coding: { code: DESC } } }, { name: DESC } ]) {
                        id
                        name
                        type {
                            coding {
                                system
                                code
                            }
                        }
                        address(sort: [ { city: ASC } ]) {
                            city
                            state
                            postalCode
                            country
                        }
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itWillApplySortSpecifiedOnInnerFieldWhenSortingByObjectField() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(sort: [ { address: { city: DESC } } ]) {
                        id
                        name
                        type {
                            coding {
                                system
                                code
                            }
                        }
                        address(sort: [ { city: DESC } ] ) {
                            city
                            state
                            postalCode
                            country
                        }
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanFilterByScalarFields() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(filter: [ { name_STARTS_WITH: "L" } ]) {
                        id
                        name
                        type {
                            coding {
                                system
                                code
                            }
                        }
                        address(filter: [ { city_ENDS_WITH: "c" } ]) {
                            city
                            state
                            postalCode
                            country
                        }
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanFilterByListScalarFields() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(sort: [ { name: ASC } ]) {
                        name
                        address(filter: [ { line_ANY_MATCH: { EQUALS: "Dlouhá 21" } } ]) {
                            city
                            line
                        }
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanFilterByCompositeFilter() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(
                        filter: [ {
                            OR: [
                                { name_STARTS_WITH: "L" },
                                { name_STARTS_WITH: "P" }
                            ]
                        } ],
                        sort: [ { name: ASC } ]
                    ) {
                        id
                        name
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanFilterByNotFilter() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(
                        filter: [ {
                            NOT: { name_STARTS_WITH: "L" }
                        } ],
                        sort: [ { name: ASC } ]
                    ) {
                        id
                        name
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanFilterByDeepObjectFields() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(sort: [ { name: ASC } ]) {
                        name
                        address(filter: [ { position: { latitude_GREATER_THAN: 50.1 } } ]) {
                            city
                            position {
                                latitude
                                longitude
                            }
                        }
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanFilterByDeepListFields() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                    findOrganizationList(filter: [ { address_ALL_MATCH: { line_ANY_MATCH: {ENDS_WITH: "1"} } } ]) {
                        name
                        address(sort: [ { city: ASC } ]) {
                            city
                            line
                        }
                    }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanFilterVeryDeeply() {
    var graphQlRequest = new GraphQlOperation(
        "",
        """
                query {
                  findStructureDefinitionList(
                    filter: [
                      {
                        differential: {
                          element_ANY_MATCH: {
                            type_ANY_MATCH: {
                              code_EQUALS: "Duration"
                            }
                          }
                        }
                      }
                    ],
                    sort: [ { type: ASC } ]
                  ) {
                    type
                    kind
                    differential {
                      element(filter: [ { type_ANY_MATCH: { code_EQUALS: "Duration"} } ], sort: [ { path: ASC } ]) {
                        path
                        type(sort: [ { code: ASC } ]) {
                          code
                        }
                      }
                    }
                  }
                }
            """
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

  @Test
  void itCanResolveVariables() {
    var graphQlRequest = new GraphQlOperation(
        "ExampleOperation",
        """
                query ExampleOperation($structureDefinitionId: String!) {
                    getStructureDefinition(id: $structureDefinitionId) {
                        id
                        kind
                        abstract
                    }
                }
            """,
        new HashMap<>(Map.of(
            "structureDefinitionId", "CodeableConcept"
        ))
    );

    var response = this.graphQlExecutor.execute(graphQlRequest);
    this.thenObjectApproved(response);
  }

//    @Test
//    public void itCanGetStructureDefinitionAndItsElements() {
//        var graphQlRequest = new GraphQLQuery(
//            "",
//            """
//                    query {
//                        getStructureDefinition(id: "Patient") {
//                            id
//                            kind
//                            abstract
//                            differential {
//                                element(
//                                    pagination: {limit: 100, offset: 0},
//                                    sort: [
//                                        { path: ASC },
//                                        {
//                                            complexUnion: {
//                                                onComplex1: {uglyName: ASC},
//                                                onComplex2: {prettyName: ASC },
//                                                onBoxedString: {value: ASC }
//                                            }
//                                        }
//                                    ]
//                                ) {
//                                    id
//                                    path
//                                    min
//                                    max
//                                    complexUnion {
//                                        ... on Complex1 {
//                                            uglyName
//                                        }
//                                        ... on Complex2 {
//                                            prettyName
//                                        }
//                                    }
//                                }
//                            }
//                        }
//                    }
//                """
//        );
//
//        var response = this.whenQueryIsDispatched(graphQlRequest, LinkedHashMap.class);
//        this.thenObjectApproved(response);
//    }
}