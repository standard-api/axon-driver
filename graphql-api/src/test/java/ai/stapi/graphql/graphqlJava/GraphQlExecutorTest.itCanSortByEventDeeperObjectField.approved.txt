{
  "findOrganizationList" : [ {
    "id" : "karlovyVaryRehabilitationCenter",
    "name" : "Karlovy Vary Rehabilitation Center",
    "address" : [ {
      "city" : "Karlovy Vary",
      "country" : "Czech Republic",
      "postalCode" : "36001",
      "state" : "Karlovarsky kraj"
    } ],
    "type" : [ {
      "coding" : [ {
        "code" : "rehabilitation-center",
        "system" : "http://terminology.hl7.org/CodeSystem/organization-type"
      } ]
    } ]
  }, {
    "id" : "ostravaPhysicalTherapy",
    "name" : "Ostrava Physical Therapy",
    "address" : [ {
      "city" : "Ostrava",
      "country" : "Czech Republic",
      "postalCode" : "70030",
      "state" : "Moravskoslezsky kraj"
    } ],
    "type" : [ {
      "coding" : [ {
        "code" : "physical-therapy",
        "system" : "http://terminology.hl7.org/CodeSystem/organization-type"
      } ]
    } ]
  }, {
    "id" : "litomericePhysicalTherapy",
    "name" : "Litomerice Physical Therapy",
    "address" : [ {
      "city" : "Litomerice",
      "country" : "Czech Republic",
      "postalCode" : "41201",
      "state" : "Ustecky kraj"
    } ],
    "type" : [ {
      "coding" : [ {
        "code" : "physical-therapy",
        "system" : "http://terminology.hl7.org/CodeSystem/organization-type"
      } ]
    } ]
  }, {
    "id" : "liberecOrthopedicClinic",
    "name" : "Liberec Orthopedic Clinic",
    "address" : [ {
      "city" : "Liberec",
      "country" : "Czech Republic",
      "postalCode" : "46001",
      "state" : "Liberecky kraj"
    } ],
    "type" : [ {
      "coding" : [ {
        "code" : "orthopedic-clinic",
        "system" : "http://terminology.hl7.org/CodeSystem/organization-type"
      } ]
    } ]
  }, {
    "id" : "ceskeBudejoviceOptometry",
    "name" : "Ceske Budejovice Optometry",
    "address" : [ {
      "city" : "Ceske Budejovice",
      "country" : "Czech Republic",
      "postalCode" : "37001",
      "state" : "Jihocesky kraj"
    } ],
    "type" : [ {
      "coding" : [ {
        "code" : "optometry",
        "system" : "http://terminology.hl7.org/CodeSystem/organization-type"
      } ]
    } ]
  }, {
    "id" : "pragueMedicalClinic",
    "name" : "Prague Medical Clinic",
    "address" : [ {
      "city" : "Aš",
      "country" : "Czech Republic",
      "postalCode" : "35201",
      "state" : "Aš"
    }, {
      "city" : "Prague",
      "country" : "Czech Republic",
      "postalCode" : "11000",
      "state" : "Praha"
    } ],
    "type" : [ {
      "coding" : [ {
        "code" : "medical-clinic",
        "system" : "http://terminology.hl7.org/CodeSystem/organization-type"
      } ]
    } ]
  }, {
    "id" : "plzenChildrensHospital",
    "name" : "Plzen Children's Hospital",
    "address" : [ {
      "city" : "Plzen",
      "country" : "Czech Republic",
      "postalCode" : "32600",
      "state" : "Plzensky kraj"
    } ],
    "type" : [ {
      "coding" : [ {
        "code" : "hospital",
        "system" : "http://terminology.hl7.org/CodeSystem/organization-type"
      } ]
    } ]
  }, {
    "id" : "olomoucHospital",
    "name" : "Olomouc Hospital",
    "address" : [ {
      "city" : "Olomouc",
      "country" : "Czech Republic",
      "postalCode" : "77200",
      "state" : "Olomoucky kraj"
    } ],
    "type" : [ {
      "coding" : [ {
        "code" : "hospital",
        "system" : "http://terminology.hl7.org/CodeSystem/organization-type"
      } ]
    } ]
  }, {
    "id" : "brnoDentalCare",
    "name" : "Brno Dental Care",
    "address" : [ {
      "city" : "Brno",
      "country" : "Czech Republic",
      "postalCode" : "60200",
      "state" : "Jihomoravsky kraj"
    } ],
    "type" : [ {
      "coding" : [ {
        "code" : "dental-care",
        "system" : "http://terminology.hl7.org/CodeSystem/organization-type"
      } ]
    } ]
  }, {
    "id" : "hradecKraloveCardiology",
    "name" : "Hradec Kralove Cardiology",
    "address" : [ {
      "city" : "Hradec Kralove",
      "country" : "Czech Republic",
      "postalCode" : "50003",
      "state" : "Kralovehradecky kraj"
    } ],
    "type" : [ {
      "coding" : [ {
        "code" : "cardiology",
        "system" : "http://terminology.hl7.org/CodeSystem/organization-type"
      } ]
    } ]
  } ]
}